work_queue/src/batch_job_sandbox.c

```diff
@@ -0,0 +1,210 @@
+#include "batch_job.h"
+#include "batch_job_internal.h"
+#include "debug.h"
+#include "process.h"
+#include "macros.h"
+#include "stringtools.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <signal.h>
+#include <ctype.h>
+
+#define CHAR_BUFF_LEN 4096
+
+static struct nvpair *output_files_list;
+static struct nvpair *task_sandbox_name_list;
+
+static batch_job_id_t batch_job_sandbox_submit (struct batch_queue *q, const char *cmd, const char *extra_input_files, const char *extra_output_files, struct nvpair *envlist )
+{
+	batch_job_id_t jobid;
+
+    // create task sandbox
+	const char *local_task_dir = nvpair_lookup_string(envlist, "local_task_dir");
+	char *template = string_format("%s/task.sandbox.XXXXXX", local_task_dir);
+    char *task_sandbox_name = mkdtemp(template);
+
+    if (output_files_list == NULL)
+    	output_files_list = nvpair_create();
+    if (task_sandbox_name_list == NULL) 
+        task_sandbox_name_list = nvpair_create();
+
+	fflush(NULL);
+	jobid = fork();
+
+	if(jobid > 0) {
+
```

Comment:
You have a concurrency problem here.  After fork(), the parent and the child are now running concurrently.  Then, you are doing the setup of the input sandbox in the *parent* process *after* the fork. That might happen to work by accident if the child process is slow to start, but it will certainly break sooner or later.  The sandbox setup needs to be complete *before* the child process is created.